pangeo_forge_recipes/executors/beam.py

```diff
@@ -0,0 +1,79 @@
+import itertools
+from dataclasses import dataclass
+from typing import Any, Iterable, List, Tuple, cast
+
+import apache_beam as beam
+
+from .base import Config, NoArgumentStageFunction, Pipeline, PipelineExecutor, Stage
+
+
+def _no_arg_stage(last: int, *, current: int, fun: NoArgumentStageFunction, config: Config) -> int:
+    """Execute a NoArgumentStageFunction, ensuring execution order."""
+    assert (last + 1) == current, f"stages are executing out of order! On step {current!r}."
+
+    fun(config=config)
+
+    return current
+
+
+def _no_op(arg, config=None) -> None:
+    pass
+
+
+@dataclass()
+class _SingleArgumentStage(beam.PTransform):
+    """Execute mappable stage in parallel."""
+
+    step: int
+    stage: Stage
+    config: Config
+
+    def prepare_stage(self, last: int) -> Iterable[Tuple[int, Any]]:
+        """Propagate current stage to Mappables for parallel execution."""
+        assert (last + 1) == self.step, f"stages are executing out of order! On step {self.step!r}."
+        return zip(itertools.repeat(self.step), cast(Iterable, self.stage.mappable))
+
+    def exec_stage(self, last: int, arg: Any) -> int:
+        """Execute stage function."""
+        assert last == self.step, f"stages are executing out of order! On step {self.step!r}."
+
+        self.stage.function(arg, config=self.config)  # type: ignore
+
+        return self.step
+
+    def post_validate(self, last: List[int]) -> int:
+        """Propagate step number for downstream stage validation."""
+        in_current_step = all((it == self.step for it in last))
+        assert in_current_step, f"stages are executing out of order! On step {self.step!r}."
+
+        return self.step
+
+    def expand(self, pcoll):
+        return (
+            pcoll
+            | "Prepare" >> beam.FlatMap(self.prepare_stage)
+            | "Execute" >> beam.MapTuple(self.exec_stage)
```

Comment:
@shoyer added a `shuffle` step here in rechunker for reasons I don't fully understand

https://github.com/pangeo-data/rechunker/blob/73ef80b4ec9dfb980988460f7d5290076d4835e2/rechunker/executors/beam.py#L69-L71

Reply:
I understand why -- and this is a good suggestion. The main reason is that it helps tell tells beam not to fuse operations, which allows the autoscale algorithm to split the work across multiple workers.